src/clj/puppetlabs/puppetserver/ringutils.clj

```diff
@@ -85,6 +109,28 @@
         false))
     true))
 
+(schema/defn ^:always-validate request->name :- (schema/maybe schema/Str)
+  "Return the name of the client by parsing the subject of the client cert from
+  the request.  Return nil if the client cert is absent or cannot be parsed."
+  [request :- RingRequest]
+  (let [{:keys [ssl-client-cert]} request]
+    (if ssl-client-cert (get-cert-subject ssl-client-cert))))
+
+(schema/defn ^:always-validate authorized? :- schema/Bool
+  "(SERVER-768) replacement for client-allowed-access?"
+  [rules :- rules/Rules
+   request :- RingRequest]
+  (if-let [name (request->name request)]
```

Comment:
as I put in [the mailing list thread](https://groups.google.com/d/msg/puppet-dev/P6vas3tvUzM/L7Gs1MvPGnIJ), I'm leaning pretty strongly towards the IP-based auth being both overly complicated and not necessary. It's more prone to spoofing, as you note, and every use-case seems to be better satisfied by having better access to certificate data. 

That said, I agree there should be a construct to permit requests with no client cert presented; at a minimum the retrieval of freshly-signed certs by nodes (though itself problematic) needs this kind of access.